File: Project 1/TicTacToe Java/src/State.java
/*
 * Sam Wlody
 * CSC 242 - Project 1
 * I affirm that I have not given or received any unauthorized
 * help on this assignment, and that this work is my own.
 */

public class State {
	// Board represented by an array - 0 is empty, 1 is "x", 2 is "o"
	private char board[];
	private char player;
	private int moveNumber;
	
	public void setTurn(char player) {
		this.player = player;
	}
	
	public char whoseTurn() {
		return this.player;
	}
	
	public State(int moveNumber) {
		this.board = new char[9];
		this.moveNumber = moveNumber;
	}
	
	public int getMoveNumber() {
		return moveNumber;
	}
	
	public char[] getBoard() {
		return board;
	}
	
	public void setBoard(char[] board) {
		this.board = board;
	}
	
	public void move(int square) {
		board[square] = player;
		player = (player == 'x' ? 'o' : 'x');
		moveNumber++;
	}
	
	public boolean checkMove(int move) {
		return board[move] != 'x' && board[move] != 'o';
	}
	
	public void print() {
		System.out.println((board[0]==0 ? '1' : board[0]) + " | " + (board[1]==0 ? '2' : board[1]) + " | " + (board[2]==0 ? '3' : board[2]));
		System.out.println("---------");
		System.out.println((board[3]==0 ? '4' : board[3]) + " | " + (board[4]==0 ? '5' : board[4]) + " | " + (board[5]==0 ? '6' : board[5]));
		System.out.println("---------");
		System.out.println((board[6]==0 ? '7' : board[6]) + " | " + (board[7]==0 ? '8' : board[7]) + " | " + (board[8]==0 ? '9' : board[8]));
	}
	
	public char getWinner() {
		if(board[0] != 0 && board[0] == board[1] && board[1] == board[2]) return board[0];
		else if(board[3] != 0 && board[3] == board[4] && board[4] == board[5]) return board[3];
		else if(board[6] != 0 && board[6] == board[7] && board[7] == board[8]) return board[6];
		else if(board[0] != 0 && board[0] == board[3] && board[3] == board[6]) return board[0];
		else if(board[1] != 0 && board[1] == board[4] && board[4] == board[7]) return board[1];
		else if(board[2] != 0 && board[2] == board[5] && board[5] == board[8]) return board[2];
		else if(board[0] != 0 && board[0] == board[4] && board[4] == board[8]) return board[0];
		else if(board[2] != 0 && board[2] == board[4] && board[4] == board[6]) return board[2];
		else if(moveNumber == 9) return 't';
		else return 'n';
	}
}

File: Project 1/TicTacToe/State.c
/*
 * Sam Wlody
 * CSC 242 - Project 1
 * I affirm that I have not given or received any unauthorized
 * help on this assignment, and that this work is my own.
 */

#include "State.h"
#include <stdlib.h>
#include <stdio.h>

State* State_init() {
    // Initial state with empty board, moveNumber 0, player 0 (x)
    State* state = (State*)malloc(sizeof(State));
    state->player = 0;
    state->moveNumber = 0;
    int i;
    // Initialize each space to -1 (representing empty)
    for(i=0; i<9; i++) {
        state->board[i] = -1;
    }
    return state;
}

State* State_next(State* state, int i) {
    // Return subsequent state of input state given a legal move
    State* nextState = (State*)malloc(sizeof(State));
    // Swap current player from input state
    nextState->player = !state->player;
    // Increment move number from input state
    nextState->moveNumber = state->moveNumber + 1;
    int m;
    // Copy board from input state
    for(m=0; m<9; m++) {
        nextState->board[m] = state->board[m];
    }
    // Assume move is legal and move input state player to space on board
    nextState->board[i] = state->player;
    return nextState;
}

void State_free(State* state) {
    free(state);
}

char State_winner(State* state) {
    // Return 1 if winning state, 0 for tie, -1 for game not yet over
    // First check to make sure first of checked squares is not empty (board[i] != -1)
    // Then check to make sure 3 in a row are the same (still not negative one)
    if(state->board[0] != -1 && state->board[0] == state->board[1] && state->board[1] == state->board[2]) return 1;
    if(state->board[3] != -1 && state->board[3] == state->board[4] && state->board[4] == state->board[5]) return 1;
    if(state->board[6] != -1 && state->board[6] == state->board[7] && state->board[7] == state->board[8]) return 1;
    if(state->board[0] != -1 && state->board[0] == state->board[3] && state->board[3] == state->board[6]) return 1;
    if(state->board[1] != -1 && state->board[1] == state->board[4] && state->board[4] == state->board[7]) return 1;
    if(state->board[2] != -1 && state->board[2] == state->board[5] && state->board[5] == state->board[8]) return 1;
    if(state->board[0] != -1 && state->board[0] == state->board[4] && state->board[4] == state->board[8]) return 1;
    if(state->board[2] != -1 && state->board[2] == state->board[4] && state->board[4] == state->board[6]) return 1;
    // If all the checks failed and is the final move, declare a tie
    if(state->moveNumber == 9) return 0;
    return -1;
}

void State_print(State* state) {
    // state->board[0] == -1 ? '1' : (state->board[0] ? 'o' : 'x')
    // If board is empty (print number) else (if x print x else print o)
    // Relies on state->board[0] return 1 (true) for o and 0 (false) for x
    printf("%c%s%c%s%c\n", state->board[0] == -1 ? '1' : (state->board[0] ? 'o' : 'x'), " | ", state->board[1] == -1 ? '2' : (state->board[1] ? 'o' : 'x'), " | ", state->board[2] == -1 ? '3' : (state->board[2] ? 'o' : 'x'));
    printf("---------\n");
    printf("%c%s%c%s%c\n", state->board[3] == -1 ? '4' : (state->board[3] ? 'o' : 'x'), " | ", state->board[4] == -1 ? '5' : (state->board[4] ? 'o' : 'x'), " | ", state->board[5] == -1 ? '6' : (state->board[5] ? 'o' : 'x'));
    printf("---------\n");
    printf("%c%s%c%s%c\n", state->board[6] == -1 ? '7' : (state->board[6] ? 'o' : 'x'), " | ", state->board[7] == -1 ? '8' : (state->board[7] ? 'o' : 'x'), " | ", state->board[8] == -1 ? '9' : (state->board[8] ? 'o' : 'x'));
}
